kale/task.py

```diff
@@ -295,3 +355,57 @@ def _report_permanent_failure(cls, message, exception, failure_type,
                        (failure_type, message.task_name, message.task_id,
                         exception))
         logger.error(message_str)
+
+
+class RemoteException(object):
+    """Utility class for reraising an exception from a remote process.
+
+    Keeps the original stack trace.
+    """
+    def __init__(self):
+        self.exc_info = sys.exc_info()
+
+    def reraise(self):
+        six.reraise(*self.exc_info)
+
+
+def _task_wrapper(task_inst, args, kwargs):
+    """Task wrapper function.
+
+    In single-process mode, this will return any exception raised.
+
+    In multi-process mode, this will return a RemoteException for any exceptions raised.
+    """
+    # This enables tasks to be blacklisted by ID or type.
+    task_inst._setup_task_environment()
+    task_inst._pre_run(*args, **kwargs)
+
+    if settings.is_multiprocess_mode():
+        # Remove SIGTERM signal handler; the parent process handles all the cleanup.
+        signal.signal(signal.SIGTERM, signal.SIG_DFL)
+
+    try:
+        # This raises an exception if the task should not be attempted.
+        task_inst._check_blacklist(*args, **kwargs)
+        task_inst.run_task(*args, **kwargs)
+    except Exception as exc:
+        # Cleanup the environment this task was running in right away.
+        task_inst._clean_task_environment(task_id=task_inst.task_id,
+                                          task_name=task_inst.task_name, exc=exc)
+        if settings.is_multiprocess_mode():
+            return RemoteException()
+        return exc
```

Comment:
hmm, it's important that we preserve the full stack trace for each task; otherwise, when a task fails, we'll have little visibility of what's going on and we'll be confused whether it's because of task code or taskworker/kale infra code.

Could we do inter-process communication to pass the exception from child process? or pass a string from child process & reconstruct the exception on parent process?